modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/CreateTableCommand.java

```diff
@@ -0,0 +1,269 @@
+            }
+
+            if (!primaryKeyColumnsSet.add(name)) {
+                throw new CatalogValidationException(format("PK column '{}' specified more that once", name));
+            }
+        }
+
+        if (nullOrEmpty(colocationColumns)) {
+            throw new CatalogValidationException("Colocation columns could not be empty");
+        }
+
+        Set<String> colocationColumnsSet = new HashSet<>();
+
+        for (String name : colocationColumns) {
+            if (!primaryKeyColumnsSet.contains(name)) {
+                throw new CatalogValidationException(format("Colocation column '{}' is not part of PK", name));
+            }
+
+            if (!colocationColumnsSet.add(name)) {
+                throw new CatalogValidationException(format("Colocation column '{}' specified more that once", name));
+            }
+        }
+    }
+
+    /**
+     * Implementation of {@link CreateTableCommandBuilder}.
+     */
+    public static class Builder implements CreateTableCommandBuilder {
+        private @Nullable List<ColumnParams> columns;
+
+        private @Nullable String schemaName;
+
+        private @Nullable String tableName;
+
+        private @Nullable List<String> primaryKeyColumns;
+
+        private @Nullable List<String> colocationColumns;
+
+        private @Nullable String zoneName;
+
+        @Override
+        public CreateTableCommandBuilder schemaName(String schemaName) {
+            this.schemaName = schemaName;
+
+            return this;
+        }
+
+        @Override
+        public CreateTableCommandBuilder tableName(String tableName) {
+            this.tableName = tableName;
+
+            return this;
+        }
+
+        @Override
+        public CreateTableCommandBuilder columns(List<ColumnParams> columns) {
+            this.columns = columns;
+
+            return this;
+        }
+
+        @Override
+        public CreateTableCommandBuilder primaryKeyColumns(List<String> primaryKeyColumns) {
+            this.primaryKeyColumns = primaryKeyColumns;
+
+            return this;
+        }
+
+        @Override
+        public CreateTableCommandBuilder colocationColumns(List<String> colocationColumns) {
+            this.colocationColumns = colocationColumns;
+
+            return this;
+        }
+
+        @Override
+        public CreateTableCommandBuilder zone(String zoneName) {
+            this.zoneName = zoneName;
+
+            return this;
+        }
+
+        @SuppressWarnings("DataFlowIssue") // params are validated in constructor
+        @Override
+        public CatalogCommand build() {
+            String zoneName = requireNonNullElse(this.zoneName, CatalogService.DEFAULT_ZONE_NAME);
+
+            List<String> colocationColumns = this.colocationColumns != null
+                    ? this.colocationColumns
+                    : primaryKeyColumns;
+
+            return new CreateTableCommand(
```

Comment:
Let's make a defensive copy for all the collections.

Reply:
good catch! Fixed